src/pybind/decoder/decodable_matrix_pybind.cc

```diff
@@ -0,0 +1,113 @@
+// pybind/decoder/decodable_matrix_pybind.cc
+
+// Copyright 2020   Mobvoi AI Lab, Beijing, China
+//                  (author: Fangjun Kuang, Yaguang Hu, Jian Wang)
+
+// See ../../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include "decoder/decodable_matrix_pybind.h"
+
+#include "decoder/decodable-matrix.h"
+
+using namespace kaldi;
+
+namespace {
+
+void pybind_decodable_matrix_scale_mapped(py::module& m) {
+  using PyClass = DecodableMatrixScaledMapped;
+  py::class_<PyClass, DecodableInterface>(m, "DecodableMatrixScaledMapped")
+      .def(py::init<const TransitionModel&, const Matrix<BaseFloat>&,
```

Comment:
This is an issue with the underlying code, not just with your wrapping.  Any time we have a const Matrix as an arg to some kind of function or constructor, it should really be MatrixBase for greater generality.  This seems to limit the re-use of PyTorch arrays, since presumably requiring a Matrix would make it necessary to copy the data (?) Please let me know if I am right about the need for copying the data.

Reply:
I agree with you. But some APIs in kaldi accepting `const Matrix<float>*` takes the owner
ship of the passed pointer, for instance, 
https://github.com/kaldi-asr/kaldi/blob/1358f383937f99ca47bbd6acd6f28831868a6fc4/src/decoder/decodable-matrix.h#L51
will delete the passed pointer in the destructor
https://github.com/kaldi-asr/kaldi/blob/1358f383937f99ca47bbd6acd6f28831868a6fc4/src/decoder/decodable-matrix.h#L75-L77

To avoid uncessary copy, can we add a new API taking `const MatrixBase<float>&`
as input ?

Reply:
Oh, I think `const Matrix<BaseFloat>*` and `const Matrix<BaseFloat>&` will not copy data since it is a pointer/reference.

Reply:
The downside is that it cannot be constructed from a numpy array.

Reply:
Another point that I feel unnatural in kaldi is that there is no table writer for `MatrixBase`
and `VectorBase`. If we want to write a `MatrixBase` to file using a table writer, we have
to convert it to  `Matrix<float>`